design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+
+We decided to run as a Job instead of Deployment because for Milestone 1 we are concerned with delivering Conjur values only once. At this time we are not supporting updating K8s Secrets with Conjur secret across multiple runs so deploying a Pod that will run continuously in the environment of a customer is a waste of their resources. Especially because the process of retrieving secrets will only happen once, at the inital spin up of the Secrets Provider.
+
+Additionally, when using Deployment to spin up a pod and the task finishes, the will container exit and be deleted. Now that the Secrets Provider needs to sit separately, when that container exists so too will the pod. Therefore, when using Deployment, Kuberentes will try to restart the Pod. After a few attempted restarts, the pod will error in a BackOff state. Because of this, we are unable to use Deployment to spin up a pod and must use a Job.
+
+#### Flow
+
+The flow for Milestone 1 is similar to the user experience from Phase 1. For each step, we will go into more detail below.
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret ***(Milestone 1)***
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role for the Service Accounts with **get**/**update** and also **list** privileges on K8s Secrets resources ***(Milestone 1)***
+
+4. Create RoleBinding and bind the Service Accounts to the role from previous step
+
+5. Create/Deploy Secrets Providers as Jobs in deployment manifest(s) ***(Milestone 1)***
+   - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s) ***(Milestone 1)***
+
+#### Customer Experience
+
+In order to preserve restrictions on Conjur secrets, a customer can decide to either:
+
+1. Deploy 1 Secrets Provider per namespace, giving the Secrets Provider access to all secrets
+2. Deploy 1 Secrets Provider per app / (group of apps), giving the Secrets Provider only access to the secrets they have been given access to in Conjur. Should the user require further separation of duties on Conjur secrets, they will need to deploy another Secrets Provider
+
+The Secrets Provider host will resemble the following:
+
+```yaml
+- !policy
+  id: secrets
+  body:
+  - !host
+    id: secrets-provider-1
+    annotations:
+      authn-k8s/namespace: prod-namespace
+      authn-k8s/service-account: prod-sa
+      authn-k8s/authenticator-container-name: cyberark-secrets-provider
+
+  - !layer 
+
+  - !permit
+    resources: !variable super_secret
+    role: !layer secrets-accessors
+    privileges: [ read, execute ]
+ 
+- !grant
+  role: !layer secrets-accessors
+  members:
+    - !host conjur/authn-k8s/secrets-provider-1
+```
+
+The Secrets Provider manifest will be defined as the following:
+
+```yaml
+apiVersion: batch/v1
```

Comment:
what is `batch/v1`? is it supported in our server?

Reply:
what do you mean
> is it supported in our server

Reply:
@orenbm ?

Reply:
this is where we have all the k8s clients: https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/k8s_object_lookup.rb#L148

in the authentication flow we need to validate the K8s Resource. we do so by using a k8s (ruby) client and get the resource. we look for a client that has the method `get_job` (in your case) so we should verify that at least one of the current clients have this method in its API.

Reply:
I manually attempted using a Job and worked as expected with no problem of authentication